src/ecc_key.c

```diff
@@ -1194,6 +1249,89 @@ static void __ec_key_reset(struct zpc_ec_key *ec_key)
 	ec_key->refcount = 1;
 }
 
+u16 ecprotkeylen_from_pvsectype(u16 pvsectype)
+{
+	switch (pvsectype) {
+	case ZPC_EC_SECRET_ECDSA_P256:
+		return 32 + 32;
+	case ZPC_EC_SECRET_ECDSA_P384:
+		return 48 + 32;
+	case ZPC_EC_SECRET_ECDSA_P521:
+		return 80 + 32;
+	case ZPC_EC_SECRET_EDDSA_ED25519:
+		return 32 + 32;
+	case ZPC_EC_SECRET_EDDSA_ED448:
+		return 64 + 32;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+/*
+ * Verify that a given pvsecret ID is a valid ID on this system, i.e. an UV
+ * secret exists with this ID and has the expected key length.
+ */
+int ec_key_blob_is_valid_pvsecret_id(struct zpc_ec_key *ec_key, const unsigned char *id)
+{
+	struct pkey_verifykey2 io;
+	unsigned char buf[sizeof(struct uvrsecrettoken)];
+	struct uvrsecrettoken *clrtok = (struct uvrsecrettoken *)&buf;
+	int rc;
+
+	memset(buf, 0, sizeof(buf));
+	clrtok->version = TOKVER_UV_SECRET;
+	clrtok->secret_type = curve2pvsecret_type[ec_key->curve];
+	clrtok->secret_len = ecprotkeylen_from_pvsectype(clrtok->secret_type);
+	memcpy(clrtok->secret_id, id, UV_SECRET_ID_LEN);
+
+	memset(&io, 0, sizeof(io));
+	io.key = buf;
+	io.keylen = sizeof(struct uvrsecrettoken);
+
+	rc = ioctl(pkeyfd, PKEY_VERIFYKEY2, &io);
+	if (rc != 0) {
+		return ZPC_ERROR_IOCTLVERIFYKEY2;
+	}
+
+	return 0;
+}
+
+/*
+ * (Re)derive protected key from a retrievable secret ID.
+ * Caller must hold aes_key's wr lock.
+ */
+int ec_key_pvsec2prot(struct zpc_ec_key *ec_key)
+{
+	struct pkey_kblob2pkey3 io;
+	unsigned char buf[sizeof(struct uvrsecrettoken)];
+	struct uvrsecrettoken *clrtok = (struct uvrsecrettoken *)&buf;
+	int rc;
+
+	memset(buf, 0, sizeof(buf));
+	clrtok->version = TOKVER_UV_SECRET;
+	clrtok->secret_type = curve2pvsecret_type[ec_key->curve];
+	clrtok->secret_len = ecprotkeylen_from_pvsectype(clrtok->secret_type);
+	memcpy(clrtok->secret_id, ec_key->cur.sec, UV_SECRET_ID_LEN);
+
+	memset(&io, 0, sizeof(io));
+	io.key = buf;
+	io.keylen = sizeof(struct uvrsecrettoken);
+	io.apqns = NULL;
+	io.apqn_entries = 0;
+	io.pkeytype = ec_key->type;
+	io.pkeylen = sizeof(ec_key->prot.protkey);
+	io.pkey = (unsigned char *)&ec_key->prot.protkey;
+
+	rc = ioctl(pkeyfd, PKEY_KBLOB2PROTK3, &io);
+	if (rc != 0) {
+		return ZPC_ERROR_IOCTLBLOB2PROTK3;
```

Comment:
DEBUG statement here?

And again, these 2 functions are very similar (almost the same), can this be refactored?

Reply:
Added the DEBUG stmt. Still thinking about refactoring ...

Reply:
OK, also did the refactoring, see new subroutine ec_key_make_uvrsecrettoken()